src/tensorOps.hpp

```diff
@@ -22,126 +22,1990 @@
-  a( 2 ) = b( 0, 2 ) * c( 0 ) + b( 1, 2 ) * c( 1 ) + b( 2, 2 ) * c( 2 );
+  for( std::ptrdiff_t i = 0; i < M; ++i )
+  {
+    dstVector[ i ] = srcVector[ i ];
+  }
 }
 
 /**
- * @tparam A The type of the matrix @p a.
- * @brief Invert the matrix @p a and @return the determinant.
- * @param a The 3x3 matrix to invert.
+ * @brief Copy @p srcMatrix into @p dstMatrix.
+ * @tparam M The size of the first dimension of @p dstMatrix and @p srcMatrix.
+ * @tparam N The size of the second dimension of @p dstMatrix and @p srcMatrix.
+ * @tparam DST_MATRIX The type of @p dstMatrix.
+ * @tparam SRC_MATRIX The type of @p srcMatrix.
+ * @param dstMatrix The destination matrix, of size M x N.
+ * @param srcMatrix The source matrix, of size M x N.
+ * @details Performs the operation @code dstMatrix[ i ][ j ] = srcMatrix[ i ][ j ] @endcode
  */
-template< typename A >
-LVARRAY_HOST_DEVICE constexpr inline
-auto invert( A const & a )
+template< std::ptrdiff_t M, std::ptrdiff_t N, typename DST_MATRIX, typename SRC_MATRIX >
+LVARRAY_HOST_DEVICE CONSTEXPR_WITHOUT_BOUNDS_CHECK inline
+void copy( DST_MATRIX && LVARRAY_RESTRICT_REF dstMatrix,
+           SRC_MATRIX const & LVARRAY_RESTRICT_REF srcMatrix )
 {
-  LVARRAY_ASSERT_EQ( a.size( 0 ), 3 );
-  LVARRAY_ASSERT_EQ( a.size( 1 ), 3 );
+  static_assert( M > 0, "M must be greater than zero." );
+  static_assert( N > 0, "N must be greater than zero." );
+  internal::checkSizes< M, N >( dstMatrix );
+  internal::checkSizes< M, N >( srcMatrix );
 
-  auto const temp00 = a( 1, 1 ) * a( 2, 2 ) - a( 1, 2 ) * a( 2, 1 );
-  auto const temp01 = a( 0, 2 ) * a( 2, 1 ) - a( 0, 1 ) * a( 2, 2 );
-  auto const temp02 = a( 0, 1 ) * a( 1, 2 ) - a( 0, 2 ) * a( 1, 1 );
-  auto const temp10 = a( 1, 2 ) * a( 2, 0 ) - a( 1, 0 ) * a( 2, 2 );
-  auto const temp11 = a( 0, 0 ) * a( 2, 2 ) - a( 0, 2 ) * a( 2, 0 );
-  auto const temp12 = a( 0, 2 ) * a( 1, 0 ) - a( 0, 0 ) * a( 1, 2 );
-  auto const temp20 = a( 1, 0 ) * a( 2, 1 ) - a( 1, 1 ) * a( 2, 0 );
-  auto const temp21 = a( 0, 1 ) * a( 2, 0 ) - a( 0, 0 ) * a( 2, 1 );
-  auto const temp22 = a( 0, 0 ) * a( 1, 1 ) - a( 0, 1 ) * a( 1, 0 );
+  for( std::ptrdiff_t i = 0; i < M; ++i )
+  {
+    for( std::ptrdiff_t j = 0; j < N; ++j )
+    {
+      dstMatrix[ i ][ j ] = srcMatrix[ i ][ j ];
+    }
+  }
+}
 
-  auto const det = a( 0, 0 ) * temp00 + a( 1, 0 ) * temp01 + a( 2, 0 ) * temp02;
-  auto const invDet = 1.0 / det;
+/**
+ * @brief Multiply the entries of @p vector by @p scale.
+ * @tparam M The size of @p vector.
+ * @tparam VECTOR The type of @p vector.
+ * @param vector The vector to scale, of length M.
+ * @param scale The value to scale the entries of @p vector by.
+ * @details Performs the operation @code vector[ i ] *= scale @endcode
+ */
+template< std::ptrdiff_t M, typename VECTOR >
+LVARRAY_HOST_DEVICE CONSTEXPR_WITHOUT_BOUNDS_CHECK inline
+void scale( VECTOR && vector, std::remove_reference_t< decltype( vector[ 0 ] ) > const scale )
```

Comment:
If you had a `sizeHelper` that returned the total size of the vector/matrix and a helper to extract the pointer to the beginning of the data, there are a bunch of vector/matrix implementations that could be a single function....which I would prefer on paper.

scale, add, scaledCopy, scaledAdd, elementWiseMultiplication...

Reply:
Slice's in general don't have contiguous storage and you need the double loop to do `x[ i ][ j ]`.